### PR TITLE
Add commands in LMCTL to retry, rollback and cancel intent process.

### PR DESCRIPTION
**Pull request by maheshlokhande1**
An intent process invoked by lmctl can subsequently be canceled, retried, or rolled back via the CP4NA GUI or its APIs. To support the rollback case, lmctl supports the processState, of the intent processes invoked by lmctl, being Rolling Back (as a transient state) or Rolled Back (as a terminal state).

cancel, retry or rollback can also be invoked using the following lmctl commands respectively, where the ID value is the processId value of the intent process.

lmctl cancel process [OPTIONS] ID
lmctl retry process [OPTIONS] ID
lmctl rollback process [OPTIONS] ID

`(env) maheshlokhande@maheshs-mbp LMCTL1 % lmctl cancel process 2ae2c2cf-cb5b-4d70-8004-3f8870f19684   
Accepted cancelling process : 2ae2c2cf-cb5b-4d70-8004-3f8870f19684`
`(env) maheshlokhande@maheshs-mbp LMCTL1 % lmctl retry process 2ae2c2cf-cb5b-4d70-8004-3f8870f19684     
Accepted retrying process : 2ae2c2cf-cb5b-4d70-8004-3f8870f19684
`

Signed-off-by: maheshlokhande1 <mahesh.lokhande@ibm.com>

### CODE REVIEW

File: lmctl/client/api/assemblies.py [hunk 1]
```diff
@@ -81,10 +90,13 @@ def _intent_request_impl(self, intent_name: str, intent_obj: Union[Dict,
...
                                                                         UpgradeAssemblyIntent]) -> str:
+        process_intents = ["retry", "rollback", "cancel"]
         endpoint = self.intent_endpoint(intent_name)
         if isinstance(intent_obj, Intent):
             intent_obj_data = intent_obj.to_dict()
         else:
             intent_obj_data = intent_obj
         request = TNCOClientRequest(method='POST', endpoint=endpoint).add_json_body(intent_obj_data)
+        if intent_name in process_intents:
+            return self._exec_request(request)
```
Comment (dvaccarosenna) [inline lmctl/client/api/assemblies.py:101]: Has it been confirmed that retry/rollback/cancel do not return a location header?

**Review by dvaccarosenna — CHANGES_REQUESTED** [review]
Comments added. There also needs to be unit test and documentation changes related to the new commands

File: lmctl/client/api/assemblies.py [hunk 2]
```diff
@@ -67,6 +67,15 @@ def intent_scale_in(self, intent_obj: Union[Dict, ScaleAssemblyIntent]) -> str:
     def intent_heal(self, intent_obj: Union[Dict, HealAssemblyIntent]) -> str:
         return self._intent_request_impl('healAssembly', intent_obj)
 
+    def intent_retry(self, intent_obj: Union[Dict, HealAssemblyIntent]) -> str:
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/client/api/assemblies.py]: Addressed.

File: lmctl/cli/commands/processes.py [hunk 3]
```diff
@@ -78,3 +82,83 @@ def get_process(
...
+    obj["processId"] = process_id
+    response = tnco_client.assemblies.intent_retry(obj)
+    io.print(f'{accepted_process_prefix} retrying process : {process_id}')
+    return
+
+@tnco_builder.make_general_command(
+    group=rollback,
+    short_help=f'Request an intent to rollback a {tnco_builder.display_name}',
+    help_prefix=f'Request an intent to rollback a {tnco_builder.display_name}',
+    help_suffix=help_suffix,
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py]: Addressed.

File: lmctl/cli/commands/processes.py [hunk 4]
```diff
@@ -78,3 +82,83 @@ def get_process(
...
+def retry_process(
+        tnco_client: TNCOClient,
+        io: IOController,
+        process_id : Identity,
+        ):
+
+    obj = {}
+    obj["processId"] = process_id
+    response = tnco_client.assemblies.intent_retry(obj)
+    io.print(f'{accepted_process_prefix} retrying process : {process_id}')
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py]: Addressed.

File: lmctl/cli/commands/processes.py [hunk 5]
```diff
@@ -78,3 +82,83 @@ def get_process(
...
+def cancel_process(
+        tnco_client: TNCOClient,
+        io: IOController,
+        process_id : Identity,
+        ):
+
+    obj = {}
+    obj["processId"] = process_id
+    response = tnco_client.assemblies.intent_cancel(obj)
+    io.print(f'{accepted_process_prefix} cancelling process : {process_id}')
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py]: Addressed.

File: lmctl/cli/commands/processes.py [hunk 6]
```diff
@@ -78,3 +82,83 @@ def get_process(
...
+@pass_io
+def rollback_process(
+        tnco_client: TNCOClient,
+        io: IOController,
+        process_id : Identity,
+        ):
+    obj = {}
+    obj["processId"] = process_id
+    response = tnco_client.assemblies.intent_rollback(obj)
+    io.print(f'{accepted_process_prefix} rolling back process : {process_id}')
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py]: Addressed.

File: lmctl/cli/commands/processes.py [hunk 7]
```diff
@@ -78,3 +82,83 @@ def get_process(
...
+    param_name='process_id',
+    obj_attribute='brokenComponentId',
+    param_opts=['process-id']
+)
+
+@tnco_builder.make_general_command(
+    group=retry,
+    short_help=f'Request an intent to retry a {tnco_builder.display_name}',
+    help_prefix=f'Request an intent to retry a {tnco_builder.display_name}',
+    help_suffix=help_suffix,
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py]: Addressed.

File: lmctl/cli/commands/processes.py [hunk 8]
```diff
@@ -78,3 +82,83 @@ def get_process(
     if limit is not None:
         query_params['limit'] = limit
     return api.query(**query_params)
+
+accepted_process_prefix = 'Accepted'
+
+help_suffix = f'''\
+For example:
+\n\nRetry process using {tnco_builder.display_name} ID: lmctl retry {tnco_builder.singular} 6ad3327e-79df-464f-af48-3283f871584d
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py:90]: Addressed.

File: lmctl/cli/commands/actions.py [hunk 9]
```diff
@@ -52,6 +52,14 @@ def get():
 def heal():
     pass
 
+@action_group(help='Retry Assembly components')
+def retry():
+    pass
+
+@action_group(help='Rollback Assembly components')
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/actions.py]: Addressed.

File: lmctl/cli/commands/actions.py [hunk 10]
```diff
@@ -52,6 +52,14 @@ def get():
 def heal():
     pass
 
+@action_group(help='Retry Assembly components')
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/actions.py]: Addressed.

File: lmctl/client/api/assemblies.py (hunk 1 above)
Comment (maheshlokhande1) [reply lmctl/client/api/assemblies.py:104]: Posted question on channel for Aubrey and Niall 
https://ibm-cloud.slack.com/archives/C044EG83NTB/p1674555284861619

Comment (maheshlokhande1) [reply lmctl/client/api/assemblies.py:104]: Niall Confirmed that response header will not have any response header.
https://ibm-cloud.slack.com/archives/C044EG83NTB/p1674556568982089?thread_ts=1674555284.861619&cid=C044EG83NTB

File: lmctl/client/api/assemblies.py [hunk 11]
```diff
@@ -81,10 +93,13 @@ def _intent_request_impl(self, intent_name: str, intent_obj: Union[Dict,
                                                                         HealAssemblyIntent, 
                                                                         ScaleAssemblyIntent,
```
Comment (dvaccarosenna) [inline lmctl/client/api/assemblies.py:94]: Need to add `RetryAssemblyIntent` and the other intent types to the list of accepted types

File: lmctl/client/models/intents.py [hunk 12]
```diff
@@ -94,6 +94,33 @@ def __eq__(self, other):
         return (isinstance(other, DeleteAssemblyIntent) and self.assembly_name==other.assembly_name
             and self.assembly_id==other.assembly_id)
 
+class RetryAssemblyIntent(ExistingAssemblyIntent):
```
Comment (dvaccarosenna) [inline lmctl/client/models/intents.py:97]: These requests expect a `process_id`, not a `assembly_id` so should use that name in the constructors. This also means they shouldn't inherit from `ExistingAssemblyIntent`

**Review by dvaccarosenna — CHANGES_REQUESTED** [review]
Additional comment - unit tests are still missing

File: lmctl/client/api/assemblies.py [hunk 13]
```diff
@@ -53,6 +55,7 @@ def intent_create_or_upgrade(self, intent_obj: Union[Dict, CreateOrUpgradeAssemb
         return self._intent_request_impl('createOrUpgradeAssembly', intent_obj)
 
     def intent_delete(self, intent_obj: Union[Dict, DeleteAssemblyIntent]) -> str:
+        print(intent_obj)
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/client/api/assemblies.py]: Addressed

File: lmctl/cli/commands/processes.py [hunk 14]
```diff
@@ -78,3 +82,93 @@ def get_process(
...
+    param_opts=['process-id']
+)
+
+@tnco_builder.make_general_command(
+    group=retry,
+    short_help=f'Request an intent to retry a {tnco_builder.display_name}',
+    help_prefix=f'Request an intent to retry a {tnco_builder.display_name}',
+    identifiers=[id_opt],
+    help_suffix=retry_help_suffix,
+    pass_file_content=False
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py]: Addressed

File: lmctl/cli/commands/processes.py [hunk 15]
```diff
@@ -78,3 +82,93 @@ def get_process(
...
     return api.query(**query_params)
+
+accepted_process_prefix = 'Accepted -'
+
+retry_help_suffix = f'''\
+For example:
+\n\nRetry process using {tnco_builder.display_name} ID: lmctl retry {tnco_builder.singular} 6ad3327e-79df-464f-af48-3283f871584d
+'''
+
+id_opt = Identifier(
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/cli/commands/processes.py]: Addressed

File: lmctl/client/api/assemblies.py [hunk 16]
```diff
@@ -81,10 +93,13 @@ def _intent_request_impl(self, intent_name: str, intent_obj: Union[Dict,
                                                                         HealAssemblyIntent, 
                                                                         ScaleAssemblyIntent,
                                                                         UpgradeAssemblyIntent]) -> str:
+        process_intents = ["retry", "rollback", "cancel"]
```
[the remark this answers is not in the document]
Comment (maheshlokhande1) [reply lmctl/client/api/assemblies.py]: Addressed

File: lmctl/client/api/assemblies.py (hunk 11 above)
Comment (maheshlokhande1) [reply lmctl/client/api/assemblies.py]: Addressed

File: lmctl/client/models/intents.py (hunk 12 above)
Comment (maheshlokhande1) [reply lmctl/client/models/intents.py]: Addressed

**Comment by dvaccarosenna** [comment]
Thanks for addressing everything so far @maheshlokhande1, it's great progress! Not sure if you've seen my comment in the request for changes about needing to add unit tests? It appears differently to code comments so could be easily missed.

**Comment by maheshlokhande1** [comment]
> Thanks for addressing everything so far @maheshlokhande1, it's great progress! Not sure if you've seen my comment in the request for changes about needing to add unit tests? It appears differently to code comments so could be easily missed.

Hi @dvaccarosenna, Added few unit tests.

File: lmctl/client/models/intents.py [hunk 17]
```diff
@@ -94,6 +94,51 @@ def __eq__(self, other):
         return (isinstance(other, DeleteAssemblyIntent) and self.assembly_name==other.assembly_name
             and self.assembly_id==other.assembly_id)
 
+class RetryAssemblyIntent(Intent):
+
+    def __init__(self, process_id: str = None):
```
Comment (dvaccarosenna) [inline lmctl/client/models/intents.py:99]: As `process_id` is the only argument to be provided I think it should not default to `None` and we should validate it has been set in the constructor, raising a `ValueError` if not

File: tests/unit/cli/commands/test_cancel_retry_rollback.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+import shutil
+from unittest.mock import patch
+from lmctl.cli.controller import clear_global_controller
+from lmctl.cli.commands.login import login
+from lmctl.config import ConfigFinder, Config
+from lmctl.environment import EnvironmentGroup, TNCOEnvironment
+import lmctl.cli.commands.processes as process_cmds
+
+
+class TestIntentCommands(command_testing.CommandTestCase):
```
Comment (dvaccarosenna) [inline tests/unit/cli/commands/test_cancel_retry_rollback.py:13]: We can name this file `test_processes.py` and the class `TestProcessCommands` to align with other unit test

**Review by dvaccarosenna — CHANGES_REQUESTED** [review]

File: lmctl/client/models/intents.py (hunk 17 above)
Comment (maheshlokhande1) [reply lmctl/client/models/intents.py:99]: Addressed.

File: tests/unit/cli/commands/test_cancel_retry_rollback.py (hunk 18 above)
Comment (maheshlokhande1) [reply tests/unit/cli/commands/test_cancel_retry_rollback.py]: Addressed

File: lmctl/client/models/intents.py (hunk 17 above)
Comment (dvaccarosenna) [reply lmctl/client/models/intents.py:99]: `process_id = " "` would be allow to pass here. We'll need to first strip whitespace `if process_id is None or len(process_id.strip()) == 0:`

**Review by dvaccarosenna — CHANGES_REQUESTED** [review]

File: lmctl/client/models/intents.py (hunk 17 above)
Comment (maheshlokhande1) [reply lmctl/client/models/intents.py:99]: Addressed.

**Review by dvaccarosenna — APPROVED** [review]

**Pull request merged**